### PR TITLE
Update custom-builds.yaml template to UAMI

### PR DESCRIPTION
**Pull request by DannyBrito**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Update custom-builds.yaml template components to use UAMI and add az cli for binary replacement download to enable AUTH in azure storage account via UAMI. 

cc: @jeremyrickard 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by mboersma** [comment]
I assume for this to work, we need kubernetes-sigs/image-builder#1326 to merge, and then to build new CAPZ reference images for e2e to test with, right?

**Comment by CecileRobertMichon** [comment]
/retest

flake tracked in https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/4123

File: templates/test/dev/custom-builds/patches/kubeadm-bootstrap-windows-k8s-pr-binaries.yaml [hunk 1]
```diff
@@ -29,6 +48,10 @@
       kube-proxy.exe --version
     path: C:/replace-pr-binaries.ps1
     permissions: "0744"
+- op: add
+  path: /spec/template/spec/preKubeadmCommands/-
+  value:
+    powershell C:/install-az-cli.ps1
```
Comment (CecileRobertMichon) [inline templates/test/dev/custom-builds/patches/kubeadm-bootstrap-windows-k8s-pr-binaries.yaml:54]: > I assume for this to work, we need https://github.com/kubernetes-sigs/image-builder/pull/1326 to merge, and then to build new CAPZ reference images for e2e to test with, right?

@mboersma I think this covers it

**Review by CecileRobertMichon — COMMENTED** [review]
We don't have a presubmit for this template in CAPZ so I opened https://github.com/kubernetes/test-infra/pull/31045 to add one so we don't regress k/k tests

**Comment by mboersma** [comment]
/retest

Network connectivity flake AFAICT.

**Comment by CecileRobertMichon** [comment]
/retest

looks like the actual cluster build succeeded

**Comment by CecileRobertMichon** [comment]
/hold

@DannyBrito I'm seeing the following error in the custom-builds job cloud-init logs:

```
[2023-10-20 01:16:16] * installing package: kubeadm v1.29.0-alpha.2.379+88d9573c3033af
[2023-10-20 01:16:22] ERROR: 
[2023-10-20 01:16:22] You do not have the required permissions needed to perform this operation.
[2023-10-20 01:16:22] Depending on your operation, you may need to be assigned one of the following roles:
[2023-10-20 01:16:22]     "Storage Blob Data Owner"
[2023-10-20 01:16:22]     "Storage Blob Data Contributor"
[2023-10-20 01:16:22]     "Storage Blob Data Reader"
[2023-10-20 01:16:22]     "Storage Queue Data Contributor"
[2023-10-20 01:16:22]     "Storage Queue Data Reader"
[2023-10-20 01:16:22]     "Storage Table Data Contributor"
[2023-10-20 01:16:22]     "Storage Table Data Reader"
[2023-10-20 01:16:22] 
[2023-10-20 01:16:22] If you want to use the old authentication method and allow querying for the right account key, please use the "--auth-mode" parameter and "key" value.
```

https://[storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/4141/pull-cluster-api-provider-azure-conformance-custom-builds/1715164798053781504/artifacts/clusters/capz-conf-6ps1ot/machines/capz-conf-6ps1ot-md-0-57lzx-8dl6m/cloud-init-output.log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/4141/pull-cluster-api-provider-azure-conformance-custom-builds/1715164798053781504/artifacts/clusters/capz-conf-6ps1ot/machines/capz-conf-6ps1ot-md-0-57lzx-8dl6m/cloud-init-output.log)

Seems like something needs to be setup on the prow subscription managed identity

**Comment by CecileRobertMichon** [comment]
Created https://github.com/kubernetes/test-infra/pull/31106 to add a windows job

**Comment by DannyBrito** [comment]
@CecileRobertMichon, the ci custom job passed, and control-plane cloud-int [logs](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/4141/pull-cluster-api-provider-azure-conformance-custom-builds/1716527010001457152/artifacts/clusters/capz-conf-ia8g0c/machines/capz-conf-ia8g0c-control-plane-pgp5n/cloud-init-output.log) look like it was able to pull binaries after role assignation to UAMI, and kubelet version is replaced in the nodes with k8s alpha version [node-describe.txt](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/4141/pull-cluster-api-provider-azure-conformance-custom-builds/1716527010001457152/artifacts/clusters/capz-conf-ia8g0c/nodes/capz-conf-ia8g0c-control-plane-rbw96/node-describe.txt)

**Comment by CecileRobertMichon** [comment]
@DannyBrito perfect. Just waiting for a review on the test-infra PR above to add a windows job so we can test this change for windows too.

**Comment by CecileRobertMichon** [comment]
```
ERROR: Operation returned an invalid status 'The specified resource name length is not within the permissible limits.'
ErrorCode:OutOfRangeInput
```

Looks like we may be running into https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3673

**Comment by CecileRobertMichon** [comment]
Opened https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4172 to fix it

**Comment by CecileRobertMichon** [comment]
Now that https://github.com/kubernetes/test-infra/pull/31126 merged, windows job should be able to run

/test pull-cluster-api-provider-azure-windows-custom-builds

**Comment by CecileRobertMichon** [comment]
/override pull-cluster-api-provider-azure-windows--custom-builds

job was renamed to fix typo

**Comment by CecileRobertMichon** [comment]
/retest

hoping that was a flake, but let's see...

**Comment by CecileRobertMichon** [comment]
Windows custom build looks good as well

```
  Kubelet Version:            v1.29.0-alpha.2.560+af52a7052b3cf5-dirty
  Kube-Proxy Version:         v1.29.0-alpha.2.560+af52a7052b3cf5-dirty
```

/lgtm
/approve

**Pull request merged**